File: alexandria/ui-elements/src/displays/components/TextEditableCode.js
import React, { Suspense } from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractTextEditableCode from "../../../gen/displays/components/AbstractTextEditableCode";
import TextEditableCodeNotifier from "../../../gen/displays/notifiers/TextEditableCodeNotifier";
import TextEditableCodeRequester from "../../../gen/displays/requesters/TextEditableCodeRequester";
import CodeBehavior from "./behaviors/CodeBehavior";
import {CircularProgress} from "@material-ui/core";

const styles = theme => ({
	editor: {
		width: "100%",
		minHeight: "100px",
		fontSize: "12pt"
	}
});

const TextEditableCodeAce = React.lazy(() => import("./texteditablecode/TextEditableCodeAce"));

class TextEditableCode extends AbstractTextEditableCode {
	state = {
		value : this.props.value
	};

	constructor(props) {
		super(props);
		this.notifier = new TextEditableCodeNotifier(this);
		this.requester = new TextEditableCodeRequester(this);
	};

	handleChange(value) {
		if (this.timeout != null) window.clearTimeout(this.timeout);
		this.timeout = window.setTimeout(() => {
			this.requester.notifyChange(value.replace(/\+/g, "&plus;"));
		}, 1000);
	};

	render() {
		const { classes, theme } = this.props;
		const value = CodeBehavior.clean(this.state.value);
		return (
			<Suspense fallback={<div className="layout horizontal center-center" style={ {margin: "10px", height: "100%"} }><CircularProgress/></div>}>
				<div style={this.style()}><TextEditableCodeAce editable={true} language={this.props.language} theme={theme} className={classes.editor}
															   width="100%" height="100%"
															   value={value} onChange={this.handleChange.bind(this)}/></div>
			</Suspense>
		);
	};

	refresh = (value) => {
		this.setState({ "value": value });
	};
}

export default withStyles(styles, { withTheme: true })(TextEditableCode);

File: alexandria/ui/src/io/intino/alexandria/ui/displays/components/TextEditableCode.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.events.ChangeEvent;
import io.intino.alexandria.ui.displays.events.ChangeListener;

public class TextEditableCode<B extends Box> extends AbstractTextEditableCode<B> {
	private String value;
	private ChangeListener changeListener = null;

	public TextEditableCode(B box) {
		super(box);
	}

	public String value() {
		return value;
	}

	public TextEditableCode<B> value(String value) {
		this.value = value.replaceAll("&#13;", "\n").replaceAll("\\\\n", "\n");
		return this;
	}

	public void update(String value) {
		this.value = value;
		notifier.refresh(value);
	}

	public TextEditableCode<B> onChange(ChangeListener listener) {
		this.changeListener = listener;
		return this;
	}

	public void notifyChange(String value) {
		this.value = value.replaceAll("&plus;", "+");
		if (changeListener != null) changeListener.accept(new ChangeEvent(this, this.value));
	}
}